### ISSUE TITLE
Invalid merge of two map[string]interface{}

### ISSUE THREAD

**Issue by dshylov** [imdario/mergo#91]
Merge of 
```
func main() {
	map1 := map[string]interface{}{
		"first": 1,
	}
	map2 := map[string]interface{}{
		"first": map[string]interface{}{
			"second" : 2,
		},
	}

	mergo.Merge(&map1, map2)

	fmt.Println(map1)
}
```

will returns

`map[first:1]`

**Comment by dshylov** [comment]
Result with `mergo.WithOverride` is the same

**Comment by birdayz** [comment]
i've just discovered the same issue. is this intended?

**Comment by dshylov** [comment]
@imdario any updates?

**Comment by 0x1mason** [comment]
@dshylov From README `You can only merge same-type structs with exported fields initialized as zero value of their type and same-types maps.` You've got different types.

**Comment by imdario** [comment]
@0x1mason, actually, maps are more flexible. Both maps have the same type map[string]interface{} so they should count as "same-type maps" but I'll check it.

**Comment by Overbryd** [comment]
I found the same.
Triggered in my helm templating. Both are of type `map[string]interface{}`.
Helm uses [Masterminds/sprig](https://github.com/Masterminds/sprig) to provide merge functionality for their template values.

This breaks the idea and "contract" implemented here: https://github.com/Masterminds/sprig/blob/master/dict.go#L80
I will also create an issue on their side.

I whipped up an example that shows the unexpected behaviour:

```go
package main

import (
  "fmt"
  "github.com/imdario/mergo"
)

func main() {
  map1 := map[string]interface{}{
    "deep": map[string]interface{}{
      "enabled": true,
    },
    "key": "default",
  }
  map2 := map[string]interface{}{
    "deep": map[string]interface{}{
      "enabled": false,
    },
    "key": "overwritten",
  }

  mergo.Merge(&map1, map2)

  fmt.Println(map1)
}
```

#### Expected output

```
map[deep:map[enabled:false] key:overwritten]
```

##### Actual output

```console
$ go get -u github.com/imdario/mergo
map[deep:map[enabled:true] key:default]
```

**Comment by vdemeester** [comment]
@Overbryd if you use `mergo.WithOverride` in the `Merge` methods, it should work..

```
package main

import (
  "fmt"
  "github.com/imdario/mergo"
)

func main() {
  map1 := map[string]interface{}{
    "deep": map[string]interface{}{
      "enabled": true,
    },
    "key": "default",
  }
  map2 := map[string]interface{}{
    "deep": map[string]interface{}{
      "enabled": false,
    },
    "key": "overwritten",
  }

  mergo.Merge(&map1, map2)

  fmt.Println(map1)
}
```

This will print `map[deep:map[enabled:false] key:overwritten]`

**Comment by Overbryd** [comment]
@vdemeester thanks, I can confirm that `mergo.Merge(&map1, map2, mergo.WithOverride)` works as expected.

**Comment by imdario** [comment]
Solved fixing #143 by reverting PR #105
